base/compiler/typeinfer.jl

```diff
@@ -235,30 +247,26 @@ function _typeinf(interp::AbstractInterpreter, frame::InferenceState)
               frames[i].cached )
         for i in 1:length(frames) ]
     empty!(frames)
-    if may_optimize(interp)
-        for (caller, _, _) in results
-            opt = caller.src
-            if opt isa OptimizationState
-                result_type = caller.result
-                @assert !(result_type isa LimitedAccuracy)
-                optimize(interp, opt, OptimizationParams(interp), result_type)
-                finish(opt.src, interp)
-                # finish updating the result struct
-                validate_code_in_debug_mode(opt.linfo, opt.src, "optimized")
-                if opt.const_api
-                    if result_type isa Const
-                        caller.src = result_type
-                    else
-                        @assert isconstType(result_type)
-                        caller.src = Const(result_type.parameters[1])
-                    end
-                elseif opt.src.inferred
-                    caller.src = opt.src::CodeInfo # stash a copy of the code (for inlining)
+    for (caller, edges, cached) in results
+        opt = caller.src
+        if may_optimize(interp) && opt isa OptimizationState
+            result_type = caller.result
+            @assert !(result_type isa LimitedAccuracy)
+            optimize(interp, opt, OptimizationParams(interp), result_type)
+            if opt.const_api
+                # XXX: The work in ir_to_codeinf! is essentially wasted. The only reason
+                # we're doing it is so that code_llvm can return the code
```

Comment:
Did you mean `code_typed`?

also duplicate: "...return the code for the return..."

Reply:
`code_typed` doesn't really care - for that we could do this mapping in `typeinf_code`. `code_llvm` goes through the regular typeinf_ext entrypoint though and expects to be returned a CodeInfo that's optimized.